### PR TITLE
feat(server): Production End Point

### PR DESCRIPTION
**Pull request by SiuYe**
#

## Changes this PR introduces

- Production endpoints to create both bike and part
- Initialize part with a cost in development (Init.js file)
<del>- Associate price with an entity created, the selling price is 30% higher than the cost price </del>
- Allow user to set profit margin

## How to test this PR

- Test each end point, create a production-part as well as a production-bike. 
- Check the price associated with that part or bike
- Make sure that this part or bike is also created

```
{
    "type": "Part",
    "componentDetail": {
        "name": "handle_bar4222256222",
        "description": "great handle bar",
        "quality": "alluminium",
        "color": "red",
        "type": "handle_bar",
        "detail": "thicc",
        "profitMargin" : "2"
    },
    "status": "Ongoing",
    "quantity": 2,
    "startDate": "2021-01-01",
    "endDate": "2021-01-03",
    "assemblyMachine": "R2-D2",
    "note": "Parts..."
}


{
    "type": "Bike",
    "componentDetail": {
        "name": "carbon road bike",
        "description": "some_description",
        "weightAmount": 5,
        "weightType": "kg",
        "color": "blue",
        "parts": {
            "handle_bar": "606917576e554255f8b0fce9",
            "wheels" : "606917576e554255f8b0fce3",
            "chain" : "606917576e554255f8b0fce4",
            "frame" : "606917576e554255f8b0fce5",
            "pedal" : "606917576e554255f8b0fce6",
            "brakes": "606917576e554255f8b0fce7",
            "seat": "606917576e554255f8b0fce8",
            "fork": "606917576e554255f8b0fce9"
        },
        "profitMargin" : "3"
    },
    "status": "Ongoing",
    "quantity": 2,
    "startDate": "2021-01-01",
    "endDate": "2021-01-03",
    "assemblyMachine": "C-3P0",
    "note": "Bikes..."
}
```

## Notes for reviewers

- Be as critical as possible 💀

## Checklist

- [x] I renamed the title of my pull request to follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] I have added tests to cover my changes (if applicable).

### CODE REVIEW

**Comment by Mackbellemore** [comment]
Task linked: [CU-nq08fz Production page](https://app.clickup.com/t/nq08fz)

File: server/app/services/PartService.ts [hunk 1]
```diff
@@ -43,7 +44,10 @@ export class PartService {
   }
 
   public async createPart(body: IPart): Promise<IPart> {
-    await this.handleMaterialStockUpdates(body.stock, body.type);
+    const price = await this.handleMaterialStockUpdates(body.stock, body.type);
+    body.costPrice = price;
+    // 30% profit from cost price
+    body.sellingPrice = price * 1.3;
```
Comment (Mackbellemore) [inline server/app/services/PartService.ts]: I feel like margin should be set by the user in the UI

Comment (Mackbellemore) [reply server/app/services/PartService.ts]: up for debate i guess

Comment (SiuYe) [reply server/app/services/PartService.ts]: sure, that could be done.

File: server/app/services/ProductionService.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  /**
+   * Create a new production json obj
+   * @param body production json obj
+   * @returns production json obj
+   */
+  public async createProduction(body: IProduction): Promise<IProduction> {
+    if (body.type === 'Part') {
+      const partEntity = {
+        name: body.componentDetail.name,
```
Comment (Mackbellemore) [inline server/app/services/ProductionService.ts:35]:
```suggestion
        ...body.componentDetail,
        stock: body.quantity
```
can u try a little something like this

Comment (Mackbellemore) [reply server/app/services/ProductionService.ts:35]: not sure if its applicable here though

File: server/app/services/ProductionService.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        quality: body.componentDetail.quality,
```
Comment (Mackbellemore) [inline server/app/services/ProductionService.ts]: seems like you have no validation on the body of the request here, if I dont pass a component detail its gonna return a "cannot read x of undefined"

Comment (Mackbellemore) [reply server/app/services/ProductionService.ts]: either you validate the body, or you optionally chain https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: server/app/models/ProductionModel.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    description: { type: Schema.Types.String, required: false },
+    quality: { type: Schema.Types.String, required: false },
+    color: { type: Schema.Types.String, required: false },
+    parts: { type: Schema.Types.Mixed, required: false },
+    type: { type: Schema.Types.String, required: false },
+    finish: { type: Schema.Types.String, required: false },
+    grade: { type: Schema.Types.String, required: false },
+    weightAmount: { type: Schema.Types.Number, required: false },
+    weightType: { type: Schema.Types.String, required: false, enum: ['kg', 'lb'] },
+    detail: { type: Schema.Types.String, required: false },
```
Comment (Mackbellemore) [inline server/app/models/ProductionModel.ts]: this feels wrong, youre basically recreating a bike in a production here. Were storing the same info in bike repository already

Comment (SiuYe) [reply server/app/models/ProductionModel.ts]: I agree, any suggestion? I need information on either part or bike when creating a production but not both

Comment (gordonpn) [reply server/app/models/ProductionModel.ts]: Right, this was actually a suggestion I gave. At the time I couldn't think of another way for the frontend to tell the backend what kind of bike and parts we'd like to put in production. I felt like this was a quick and dirty way to do it.

Maybe we can use a reference Mongo doc ID instead?

Comment (Mackbellemore) [reply server/app/models/ProductionModel.ts]: ya the request body can contain all of this but I think we should just store the mongo ID in the production collection

Comment (Mackbellemore) [reply server/app/models/ProductionModel.ts]: or we completely remove the Bike code we have on the server

Comment (gordonpn) [reply server/app/models/ProductionModel.ts]: Oh, yeah you're 100% right. The request body needs this info but not the MongoDB document

File: server/app/services/ProductionService.ts (hunk 3 above)
Comment (gordonpn) [reply server/app/services/ProductionService.ts]: @SiuYe Using the following POST request body
```json
{
  "type": "Part"
}
```
I get a 500 error status code (unhandled error) because the code is trying to access a property of an undefined object, is that the expected behaviour?

File: server/app/models/ProductionModel.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { Document, Schema } from 'mongoose';
+
+export interface IProduction extends Document {
+  type: string;
+  componentId: string;
```
Comment (gordonpn) [inline server/app/models/ProductionModel.ts]: Is `componentId` useful to have on the request body schema?

I think it would be useful if on the frontend you are planning to allow users to select a pre-existing part/bike to produce more of, but I don't see that kind of usage for the `componentId` in the backend? Unless I missed it?

File: server/app/services/ProductionService.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        name: body.componentDetail.name,
+        description: body.componentDetail.description,
+        weightAmount: body.componentDetail.weightAmount,
+        weightType: body.componentDetail.weightType,
+        color: body.componentDetail.color,
+        parts: body.componentDetail.parts,
+        stock: body.quantity,
+        profitMargin: body.componentDetail.profitMargin,
+      } as IBike;
+      const validBike = await BikeEntity.validate(bikeEntity, 'create');
```
Comment (gordonpn) [inline server/app/services/ProductionService.ts]:
```suggestion
      const validBike = await BikeEntity.validate(
        ({ ...body.componentDetail, stock: body.quantity } as unknown) as IBike,
        'create'
      );
```

This might be cleaner? (Can do similarly with `IPart` above)

@Mackbellemore Is it equivalent?

Comment (Mackbellemore) [reply server/app/services/ProductionService.ts]: As long as componentDetail doesnt have extra stuff that a Bike doesnt have then ya something like this works.

Comment (SiuYe) [reply server/app/services/ProductionService.ts]: @gordonpn NICE! it's much cleaner I agree. 👍

File: server/app/services/ProductionService.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+   * @param body production json obj
+   * @returns production json obj
+   */
+  public async createProduction(body: IProduction): Promise<IProduction> {
+    if (body.type === 'Part') {
+      const validPart = await PartEntity.validate(
+        ({
+          ...body.componentDetail,
+          stock: body.quantity,
+          profitMargin: body.componentDetail.profitMargin,
```
Comment (gordonpn) [inline server/app/services/ProductionService.ts]: Why did you add the profitMargin? Is it not in the other object spread already?

Comment (SiuYe) [reply server/app/services/ProductionService.ts]: You are right! 😐

File: server/app/models/ProductionModel.ts (hunk 5 above)
Comment (SiuYe) [reply server/app/models/ProductionModel.ts]: right it's not necessary for the request. It was more to store the ID of the Part/Bike created. 

https://github.com/Mackbellemore/soen-390-team07/blob/eb3e920cc6bab8b645ca01ac71f5e8845cde5f66/server/app/services/ProductionService.ts#L45-L46

Do you have a suggestion on how I can remove it from the interface but still have it stored?

Comment (gordonpn) [reply server/app/models/ProductionModel.ts]: @SiuYe What's your plan with this componentId? If it's not used then remove it

**Review by Mackbellemore — COMMENTED** [review]
fix ye merge conflicts boi

File: server/app/models/ProductionModel.ts (hunk 5 above)
Comment (SiuYe) [reply server/app/models/ProductionModel.ts]: It's used to create the production. The Id is stored and used in the UI to display the entity created.

Comment (SiuYe) [reply server/app/models/ProductionModel.ts]: @gordonpn thanks for the solution & help!

**Review by gordonpn — APPROVED** [review]

**Review by Mackbellemore — APPROVED** [review]
